### PR TITLE
unified distance calls

### PR DESCRIPTION
**Pull request by nishaq503**
feature: unified distance calls into single function

### CODE REVIEW

File: pyclam/manifold.py [hunk 1]
```diff
@@ -133,13 +135,15 @@ def argsamples(self) -> Vector:
             else:
                 n = int(np.sqrt(len(self)))
                 indices = list(np.random.choice(self.argpoints, n, replace=False))
+                indices = [int(i) for i in indices]
 
             # Handle Duplicates.
-            if pdist(self.manifold.data[indices], self.metric).max(initial=0.) == 0.:
+            if self.distance(x1=indices, x2=indices).max(initial=0.) == 0.:
```
Comment (thoward27) [inline pyclam/manifold.py]: Recommendation to not use variable names here, this function *almost* justifies being position only. It doesn't matter which comes first.

File: pyclam/manifold.py [hunk 2]
```diff
@@ -133,13 +135,15 @@ def argsamples(self) -> Vector:
...
                 indices = np.unique(self.manifold.data[self.argpoints], return_index=True, axis=0)[1]
                 indices = [self.argpoints[i] for i in indices][:n]
 
             # Cache it.
+            indices = [int(i) for i in indices]
```
Comment (thoward27) [inline pyclam/manifold.py]: Why int here, they should be ints already, if they aren't we fucked up elsewhere

File: pyclam/manifold.py [hunk 3]
```diff
@@ -133,13 +135,15 @@ def argsamples(self) -> Vector:
             else:
                 n = int(np.sqrt(len(self)))
                 indices = list(np.random.choice(self.argpoints, n, replace=False))
+                indices = [int(i) for i in indices]
```
Comment (thoward27) [inline pyclam/manifold.py]: can't we do something like `random.choice(dtype=int)`

File: pyclam/manifold.py [hunk 4]
```diff
@@ -163,7 +167,7 @@ def argmedoid(self) -> int:
         """ The index used to retrieve the medoid. """
         if '_argmedoid' not in self.__dict__:
             logging.debug(f"building cache for {self}")
-            _argmedoid = np.argmin(cdist(self.samples, self.samples, self.metric).sum(axis=1))
+            _argmedoid = np.argmin(self.distance(x1=self.argsamples, x2=self.argsamples).sum(axis=1))
```
Comment (thoward27) [inline pyclam/manifold.py]: same here, would recommend against using keyword arguments for functions where the order doesn't matter

Comment (thoward27) [reply pyclam/manifold.py]: https://www.python.org/dev/peps/pep-0570/

File: pyclam/manifold.py [hunk 5]
```diff
@@ -207,7 +211,7 @@ def local_fractal_dimension(self) -> float:
                 return 0.
             count = [d <= (self.radius / 2)
                      for batch in self
-                     for d in self.distance(self.manifold.data[batch])]
+                     for d in self.distance(x1=self.argmedoid, x2=batch)[0]]
```
Comment (thoward27) [inline pyclam/manifold.py]: why add the `[0]`

Comment (thoward27) [reply pyclam/manifold.py]: this would indicate to me that the distance functions are materially different

File: pyclam/manifold.py [hunk 6]
```diff
@@ -231,14 +235,14 @@ def tree_search(self, point: Data, radius: Radius, depth: int) -> Dict['Cluster'
 
         results: Dict['Cluster', Radius] = dict()
         if self.depth == depth:
-            results = {self: self.distance(np.asarray([point]))[0]}  # TODO: Cover
+            results = {self: self.distance(x1=self.argmedoid, x2=None, p2=point)[0]}  # TODO: Cover
```
Comment (thoward27) [inline pyclam/manifold.py]: I definitely don't understand what p2 is intended to signify here

File: pyclam/manifold.py [hunk 7]
```diff
@@ -258,8 +262,8 @@ def _tree_search(self, point: Data, radius: Radius, depth: int) -> Dict['Cluster
                 break
 
             # filter out clusters that are too far away to possibly contain any hits.
-            centers = np.asarray([c.medoid for c in children])
-            distances = cdist(np.expand_dims(point, 0), centers, self.metric)[0]
+            argcenters = [c.argmedoid for c in children]
+            distances = self.distance(x1=None, p1=point, x2=argcenters)[0]
```
Comment (thoward27) [inline pyclam/manifold.py]: weird, I think this would read better if params were `x1, x2` and that's it.

File: pyclam/manifold.py [hunk 8]
```diff
@@ -405,6 +398,9 @@ def __hash__(self):
     def __contains__(self, cluster: 'Cluster') -> bool:
         return cluster in self.clusters.keys()
 
+    def distance(self, *, x1: Union[int, List[int], None], x2: Union[int, List[int], None], p1: Data = None, p2: Data = None) -> np.ndarray:
+        return self.manifold.distance(x1=x1, x2=x2, p1=p1, p2=p2)
```
Comment (thoward27) [inline pyclam/manifold.py]: Disagree with this implementation. I'd rather see just x1 x2. p1 p2 is super weird here for me

Comment (thoward27) [reply pyclam/manifold.py]: https://www.python.org/dev/peps/pep-0570/

File: pyclam/manifold.py [hunk 9]
```diff
@@ -608,6 +605,48 @@ def __repr__(self) -> str:
...
+
+        This DOES NOT do any batching.
+
+        The metric given to Manifold should have the following properties:
+            * dist(p1, p2) = 0 if and only if p1 = p2.
+            * dist(p1, p2) = dist(p2, p1)
+
+        :param x1: index of point or list of indexes of points in self.data.
+        :param x2: index of point or list of indexes of points in self.data.
+        :param p1: (optional) point(s) given directly to override x1. Useful when the point is not in self.data
```
Comment (thoward27) [inline pyclam/manifold.py]: weird, I find this p1 p2 notion unintuitive, I'd say make it work with just x1 x2 and make them positional only

File: pyclam/manifold.py [hunk 10]
```diff
@@ -608,6 +605,48 @@ def __repr__(self) -> str:
     def depth(self) -> int:
         return len(self.graphs) - 1
 
+    def distance(self, *, x1: Union[int, List[int], None], x2: Union[int, List[int], None], p1: Data = None, p2: Data = None) -> np.ndarray:
```
Comment (thoward27) [inline pyclam/manifold.py]: I see now what you intended here, to override if point not in self.data, but this function should not do that. Call it `argdistance` if you want.

Comment (thoward27) [reply pyclam/manifold.py]: perhaps implement checks on type, list of ints is indices, numpy array is a dedicated point. Something other than this param string

Comment (thoward27) [reply pyclam/manifold.py]: Just had a thought.. check dimensions. 1D arrays are indices, 2D are points. State that in the docstring, and make this positional, that would be very satisfying

File: pyclam/manifold.py (hunk 3 above)
Comment (nishaq503) [reply pyclam/manifold.py]: I think numpy just maps that to np.int from native ints

File: pyclam/manifold.py (hunk 2 above)
Comment (nishaq503) [reply pyclam/manifold.py]: Line 143 makes it a list of numpy.int64

File: pyclam/manifold.py [hunk 11]
```diff
@@ -608,6 +607,31 @@ def __repr__(self) -> str:
...
+        :param x1: a list of indices, or a 2D matrix of data points
+        :param x2: a list of indices, or a 2D matrix of data points
+        :return: matrix of pairwise distances.
+        """
+
+        x1, x2 = np.asarray(x1), np.asarray(x2)
+        
+        # Fetch data if given indices.
+        if len(x1.shape) < 2:
+            x1 = self.data[x1 if x1.ndim == 1 else np.expand_dims(x1, 0)]
```
Comment (nishaq503) [inline pyclam/manifold.py]: This is wrong.
When x1.shape is less than 2, x1.ndim is always 1

Comment (thoward27) [reply pyclam/manifold.py]: No, it can be zero. `len(np.array(1))`

File: pyclam/tests/test_cluster.py [hunk 12]
```diff
@@ -182,7 +180,7 @@ def test_neighbors(self):
         return
 
     def test_distance(self):
-        self.assertGreater(self.children[0].distance(np.expand_dims(self.children[1].medoid, 0)), 0)
+        self.assertGreater(self.children[0].distance(x1=self.children[0].argmedoid, x2=self.children[1].argmedoid), 0)
```
Comment (thoward27) [inline pyclam/tests/test_cluster.py]: Drop arguments here

**Pull request merged**